2630 색종이 만들기 파이썬 반영

hyunjinkim-kr opened the pull request:
## 문제
- 색종이 만들기
- https://www.acmicpc.net/problem/2630

## 해결책
- 가장 최소단위 사각형에 대해 판별하고, 가로세로가 2등분 되기 전의 네 곳에 대해 동일한 색인지 확인한다.

### 세부 사항
1. **변수 설명**:
    - N : 정사각형 색종이의 한 변 길이
    - array : 색종이의 형태가 저장되는 2차원배열
    - x : 행
    - y : 열
    - size : 탐색하는 종이의 범위
    - positions : 4부분의 위치를 저장하는 배열
    - white : 흰색 (0) 정사각형의 개수
    - blue : 파란색 (1) 정사각형의 개수

## 알고리즘 설명
 1. N 과 N 길이의 색종이 내용을 입력받아 array에 저장한다.
 2. 색종이 사이즈가 1인경우, x,y에 해당하는 array의 숫자가 0이면 (1,0)을, 1이면 (0,1)을 리턴한다. 사이즈가 1이 아닌경우, 아래 과정을 반복한다.
  2-1 사이즈를 2로 나눈후, 초기 크기의 4등분에 해당하는 positions의 사각형에 대해 재귀하며 계산한다.
  2-2 리턴된 blue 값이 0이면 흰색 (1,0), white 값이 0이면 파란색 (0,1)을 리턴한다.
  2-3 모두 같지 않은경우 (white,blue)를 리턴한다.
 3. white 와 blue를 출력한다.

## 시간복잡도
- $O(N^2)$

hyunjinkim-kr committed:
2630 색종이 만들기 파이썬 반영

kyeongmincho commented on python-algorithm/step-by-step/divideConquer/solve_2630.py:
```diff
@@ -0,0 +1,54 @@
+import sys
+
```
이전과 마찬가지로, 소스코드에도 해결책을 간단히 적어주셈

kyeongmincho commented on python-algorithm/step-by-step/divideConquer/solve_2630.py:
```diff
@@ -0,0 +1,54 @@
+    # n줄에 걸쳐서 2차원 리스트 입력 받기
+    array = []
+    for _ in range(n):
+        array.append(list(map(int, sys.stdin.readline().split())))
```
넘 복잡... 좀 분리해주시길..

kyeongmincho commented on python-algorithm/step-by-step/divideConquer/solve_2630.py:
```diff
@@ -0,0 +1,54 @@
+
+# 분할정복 함수
+def solve(array, x, y, size):
+    global white, blue
```
웬만하면 전역변수는 피하자

kyeongmincho commented on python-algorithm/step-by-step/divideConquer/solve_2630.py:
```diff
@@ -0,0 +1,54 @@
+    global white, blue
+
+    # 현재 영역이 모두 같은 값인지 확인하는 함수
+    if is_uniform(array, x, y, size):
```
전역변수를 사용하지 않기로 하는 순간
- 왜 여기에 if/else 문이 필요 없어지고
- 더 좋은 시간복잡도로 풀리는 게 가능한지 알게 될 듯?

kyeongmincho commented on python-algorithm/step-by-step/divideConquer/solve_2630.py:
```diff
@@ -0,0 +1,54 @@
+        array.append(list(map(int, sys.stdin.readline().split())))
+
+    # 문제 해결
+    solve(array, 0, 0, n)
```
- "0, 0," 같은 건 사용자 친화적인 파라미터가 아님. 경우에 따라 wrapper 함수를 써야한다는 것을 예전에 배운 적이 있음
- (그러나 아래 코멘트를 참고로 시간복잡도를 줄이게 된다면 결국 어찌됐건 코드가 바뀌긴 할 듯)

kyeongmincho reviewed (changes requested):
> O(N^2)

- 시간복잡도 더 줄일 수 있을 것임. (힌트: 일단 쪼개기..)
- 참고로 깃헙에서 latex 문법으로 수식 표현할 수 있음 $O(N^2)$(`$O(N^2)$`) 처럼 $로 감싸면 됨. ([관련 문서](https://docs.github.com/ko/get-started/writing-on-github/working-with-advanced-formatting/writing-mathematical-expressions))

hyunjinkim-kr committed:
1920 수 찾기 자바 수정중

hyunjinkim-kr committed:
2630 색종이 만들기 파이썬 수정 :
is_uniform삭제, bottom-up 방식 수정, 1개 단위까지 쪼갠 후 머지하여 return

iyaacs reviewed (commented):
LGTM (따봉)

kyeongmincho commented on python-algorithm/step-by-step/divideConquer/solve_2630.py:
```diff
@@ -0,0 +1,82 @@
+# 2. 색종이 사이즈가 1인경우, x,y에 해당하는 array의 숫자가 0이면 [1,0]을, 1이면 [0,1]을 리턴한다. 사이즈가 1이 아닌경우, 아래 과정을 반복한다.
+#  2-1 사이즈를 2로 나눈후, 초기 크기의 4등분 크기의 사각형에 대해 재귀하며 계산한다.
+#  2-2 각 사각형의 결과가 모두 같은경우, 리턴된 배열의 0번째 값이 1이면 흰색, 1번째 값이 1이면 파란색을 리턴한다.
+#    e.g) 0일떄 return [1,0]
```
"e.g."

hyunjinkim-kr replied:
죄송

kyeongmincho commented on python-algorithm/step-by-step/divideConquer/solve_2630.py:
```diff
@@ -0,0 +1,82 @@
+            return [0, 1]
+
+    # 그렇지 않으면 그대로 합산된 결과를 반환
+    return [white, blue]
```
틀린 건 아니지만, 이 문제의 경우 반환 타입이 값 "두" 개로 고정이니까 (즉, 가변 길이가 아니니까), 반환 타입을 list보다는 tuple이 어떨지? (i.e. `(white, blue)`)

hyunjinkim-kr replied:
변경 완료 result = (white, blue)

kyeongmincho commented on python-algorithm/step-by-step/divideConquer/solve_2630.py:
```diff
@@ -0,0 +1,82 @@
+            return [0, 1]  # 파란색 (1)
+    # 4등분으로 나누어 재귀 호출
+    half_size = size // 2
+    result1 = solve_inner(array, x, y, half_size)  # 왼쪽 위
```
```py
postions = [(x, y), (x, y + half_size), ...]  # 타겟 포지션을 리스트에 담고
results = []
for x, y in positions:  # 이런 식으로 순회하는 게
    results = solve_inner(array, x, y ...)  # 더 깔끔하지 않을지?
```

hyunjinkim-kr replied:
반영완

kyeongmincho commented on python-algorithm/step-by-step/divideConquer/solve_2630.py:
```diff
@@ -0,0 +1,82 @@
+    result3 = solve_inner(array, x + half_size, y, half_size)  # 왼쪽 아래
+    result4 = solve_inner(array, x + half_size, y + half_size, half_size)  # 오른쪽 아래
+    # 네 부분에서 얻은 결과를 합산
+    white = result1[0] + result2[0] + result3[0] + result4[0]
+    blue = result1[1] + result2[1] + result3[1] + result4[1]
```
(위 코멘트에 이어서)
```py
white_list, blue_list = zip(*results)
white = sum(white_list)
blue = sum(blue_list)
```
혹은 (문법이 정확한지 모르겠지만...)
```py
# `from functools import reduce` 가 선언되어야 함
white, blue = reduce(lambda acc, (w, b): (acc[0] + w, acc[1] + b), results, (0, 0))
```

같은 방식으로 "higher-order function"으로 간결하게 표현할 수 있음. 나중에 만나면 가르쳐주려고 하긴 했는데, 여기서 사용된 `zip()`, `reduce()`와 더불어 `map()`, `filter()` 등에 대해 혼자서 공부하고 이해해볼래?

hyunjinkim-kr replied:
```python
for x, y in positions:
        zipped = zip(results, chk_paper_color(array, x, y, half_size))  # 다음과 같은 방식은 어떨지?
        results = tuple(map(sum, zipped))

white = results[0]
blue = results[1]

kyeongmincho commented on python-algorithm/step-by-step/divideConquer/solve_2630.py:
```diff
@@ -0,0 +1,82 @@
+def solve_inner(array, x, y, size):
+    # 현재 영역이 모두 같은 값인지 확인하는 함수
+    if size == 1:
+        if array[x][y] == 0:
+            return [1, 0]  # 하얀색 (0)
+        else:
+            return [0, 1]  # 파란색 (1)
```
아마 좀 더 pythonic한 코드는
```py
return [1, 0] if array[x][y] == 0 else [0, 1]
```
일 거야

hyunjinkim-kr replied:
```python
#해당 코드에서도 저렇게 바꿔야 하는지?
if blue == 0:  # 모두 하얀색(0)
        return (1, 0)
    elif white == 0:  # 모두 파란색(1)
        return (0, 1)

kyeongmincho replied:
??

kyeongmincho reviewed (changes requested):
문제 잘 풀었음! 고오급스러운 코딩 스타일에 대해 좀 더 공부해보자

kyeongmincho commented on python-algorithm/step-by-step/divideConquer/solve_2630.py:
```diff
@@ -0,0 +1,82 @@
+    blue = result1[1] + result2[1] + result3[1] + result4[1]
+
+    # 만약 4개의 구역이 모두 동일하다면, 그 구역을 하나로 합칠 수 있음
+    if result1 == result2 == result3 == result4:
```
`white`, `blue`까지 다 따로 구해놨는데 그냥 한쪽이 0인 경우를 체크하는 게 낫지 않을지?

hyunjinkim-kr replied:
확인 ㅎ.....

kyeongmincho commented on python-algorithm/step-by-step/divideConquer/solve_2630.py:
```diff
@@ -0,0 +1,82 @@
+
+
+# 분할정복 함수
+def solve_inner(array, x, y, size):
```
좀 더 semantic한 이름을 줘보자. `solve`까지는 약간 템플릿 같은 느낌의 이름이라 그렇다치고, solve 내부에서 사용되는 함수들은 문제와 밀접환 관련이 있으므로 좀 더 뜻이 있는 이름이 있었으면 좋겠음. 앞으로도 마찬가지..

hyunjinkim-kr replied:
```python
def chk_paper_color(array, x, y, size):

hyunjinkim-kr committed:
1920 수 찾기 수정 : 튜플() 사용, position 배열 생성 및 리턴방법 변경

kyeongmincho commented on python-algorithm/step-by-step/divideConquer/solve_2630.py:
```diff
@@ -0,0 +1,87 @@
+        zipped = zip(results, chk_paper_color(array, x, y, half_size))
+        results = tuple(map(sum, zipped))
+
+    white = results[0]
+    blue = results[1]
```
`white, blue = results`
어떤지?

kyeongmincho reviewed (approved):
남은 코멘드들 알아서 처리하고 머지하면 될 듯..
LGTM!

hyunjinkim-kr committed:
solve_2630 색종이 만들기 :
1. 표현 변경 return [1, 0] if array[x][y] == 0 else [0, 1]
2. white, blue = results
